Rethink the approach to documenting package management (#3433)
theforeman/foreman-documentation: guides/common/modules/ref_managing-packages-on-the-base-operating-system.adoc

[id="Managing_Packages_on_the_Base_Operating_System_{context}"]
= Managing packages on the base operating system of {ProjectServer} or {SmartProxyServer}

{Project} provides the `{foreman-maintain} packages` command to install and update packages on the base operating system of your {ProjectServer} or {SmartProxyServers}.

[WARNING]
====

> Installing and updating individual packages with `yum` is not supported.
> While {Project} provides the `{foreman-maintain} packages unlock` command to bypass `{foreman-maintain} packages` and enable package management with `yum`, this can lead to package inconsistencies and compromise the stability of the system.

Reviewer: ```suggestion
Installing and updating individual packages with `dnf` is not supported.
While {Project} provides the `{foreman-maintain} packages unlock` command to bypass `{foreman-maintain} packages` and enable package management with `dnf`, this can lead to configuration inconsistencies and compromise the stability of the system.
```

> Always use `{foreman-maintain} packages` to manage packages.

Reviewer: Does `manage` also mean `uninstall`? Because we don't offer a command for that :)

Author: Does that mean that if I install a package with foreman-maintain, then there is no supported way of uninstalling it? @evgeni

Reviewer: It's complicated ;-)
The dnf plugin that we ship, which is responsible for this locking mechanism, doesn't affect removals, so people can do `dnf remove blah` just fine.
This might havoc their system in other ways, but that's not what the locking feature is trying to prevent.

Author: Thanks for explaining. I've changed this to "install and update".

I'm leaving "Managing" in the heading unchanged because that heading is already too long and I don't think it's such an issue there.

====

> After you install or update a package on the base operating system, running `{foreman-installer}` afterwards is required.

Reviewer 2: ```suggestion
You must run `{foreman-installer}` after you install or update a package on the base operating system.
```

alternative:

```suggestion
You must run `{foreman-installer}` after installing or updating a package on the base operating system.
```

Author: I actually agonized over this for quite some time. Writing "you must run" is not entirely true because if you manage the packages in the supported way (`foreman-maintain packages`), `foreman-installer` is run automatically as part of `foreman-maintain packages`.

But if you manage packages in the non-supported way (unlock + dnf/yum), you might not know that `foreman-installer` is required too.

While we don't support the other way but in reality, some users do it anyway. In which case we don't want them to break their Foreman _too much_. So I was looking for a wording that would support that scenario -- while being careful not to encourage it.

While I know that the wording I used is not great, I also feel that I can't accept your suggestions.

Author: I look at the whole paragraph again and reworded it slightly to make it sound more natural. I hope it helps at least a little bit.

When you manage packages with `{foreman-maintain} packages`, the command automatically runs `{foreman-installer}` as well.

> Note that this results in restarting some services on the base operating system.

Reviewer 2: ```suggestion
Note that this results in restarting any required services on the base operating system.
```

"some" feels very unspecific. Is "any required" any better? Or maybe "restarting affected services"?

Author: I think I like "affected services", thanks!

> Note that this results in restarting affected services on the base operating system.

Reviewer 3: I think the installer doesn't manage services on the base OS but really the Satellite services.

Reviewer: How about:
```suggestion
Note that this might result in restarts of services when their configuration changes due to the updates.
```
(completely avoiding OS vs Satellite)

Author: Applied, slightly modified.

.Overview of `{foreman-maintain} packages` subcommands
[cols=2,1,options="header"]
|====
|Subcommand |Description
a|
[options="nowrap", subs="+quotes,attributes"]
----

> {project-package-install} _package_1_ _package_2_

Reviewer 2: ```suggestion
{project-package-install} _My_Package_1_ _My_Package_2_
```

Author: Applied. I'm somewhat concerned about what it looks like in the preview. I'm thinking that I could perhaps drop `My_Package_2` to make the code blocks a bit shorter. @maximiliankolb

Reviewer 2: Yeah, you could either use `_My_Package_` or `_My_Packages_` as a shorter placeholder.

Author: Thanks, let's go with `My_Package`.

----
|Installs packages on {ProjectServer} or {SmartProxyServer}
a|
[options="nowrap", subs="+quotes,attributes"]
----
{project-package-check-update}
----
|Checks for available package updates on {ProjectServer} or {SmartProxyServer}
a|
[options="nowrap", subs="+quotes,attributes"]
----
{project-package-update}
----
|Updates all packages on {ProjectServer} or {SmartProxyServer}
a|
[options="nowrap", subs="+quotes,attributes"]
----

> {project-package-update} _package_1_ _package_2_

Reviewer 2: same as above.

----
|Updates specific packages on {ProjectServer} or {SmartProxyServer}
a|
[options="nowrap", subs="+quotes,attributes"]
----
{foreman-maintain} packages lock
----

> |Reverts `{foreman-maintain} unlock` and configures {Project} to ensure the stability of the base operating system by preventing users from managing packages with `yum`

Reviewer 2: ```suggestion
|Reverts `{foreman-maintain} unlock` and configures {Project} to ensure the stability of the base operating system by preventing users from managing packages with `yum` and `dnf`.
```

> |Reverts `{foreman-maintain} unlock` and configures {Project} to ensure the stability of the base operating system by preventing users from managing packages with `dnf`.

Author: Ok, let's go with your suggestion. (This was supposed to be a hint for users who _have_ run unlock and now are trying to revert to the previous state... but I don't feel too strongly about it.)
